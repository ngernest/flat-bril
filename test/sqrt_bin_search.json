{
  "functions": [
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "min_guess",
          "type": "int"
        },
        {
          "name": "max_guess",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "min_guess"
          ],
          "dest": "v0",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v1",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v0",
            "v1"
          ],
          "dest": "v2",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v2"
          ],
          "dest": "min_plus_1",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "min_plus_1"
          ],
          "dest": "v4",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "max_guess"
          ],
          "dest": "v5",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v4",
            "v5"
          ],
          "dest": "v6",
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v6"
          ],
          "labels": [
            "then.3",
            "else.3"
          ],
          "op": "br"
        },
        {
          "label": "then.3"
        },
        {
          "args": [
            "min_guess"
          ],
          "dest": "v7",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v7"
          ],
          "op": "print"
        },
        {
          "dest": "v8",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "op": "ret"
        },
        {
          "label": "else.3"
        },
        {
          "args": [
            "min_guess"
          ],
          "dest": "v9",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "max_guess"
          ],
          "dest": "v10",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v9",
            "v10"
          ],
          "dest": "v11",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v11"
          ],
          "dest": "sum_bounds",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "sum_bounds"
          ],
          "dest": "v12",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v13",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "v12",
            "v13"
          ],
          "dest": "v14",
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "v14"
          ],
          "dest": "mid",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "mid"
          ],
          "dest": "v15",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "mid"
          ],
          "dest": "v16",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v15",
            "v16"
          ],
          "dest": "v17",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v17"
          ],
          "dest": "mid_squared",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "mid_squared"
          ],
          "dest": "v19",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v20",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v19",
            "v20"
          ],
          "dest": "v21",
          "op": "le",
          "type": "bool"
        },
        {
          "args": [
            "v21"
          ],
          "labels": [
            "then.18",
            "else.18"
          ],
          "op": "br"
        },
        {
          "label": "then.18"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v22",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "mid"
          ],
          "dest": "v23",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "max_guess"
          ],
          "dest": "v24",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v22",
            "v23",
            "v24"
          ],
          "funcs": [
            "main"
          ],
          "op": "call"
        },
        {
          "dest": "v25",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "labels": [
            "endif.18"
          ],
          "op": "jmp"
        },
        {
          "label": "else.18"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v26",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "min_guess"
          ],
          "dest": "v27",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "mid"
          ],
          "dest": "v28",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v26",
            "v27",
            "v28"
          ],
          "funcs": [
            "main"
          ],
          "op": "call"
        },
        {
          "dest": "v29",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "label": "endif.18"
        }
      ],
      "name": "main"
    }
  ]
}
