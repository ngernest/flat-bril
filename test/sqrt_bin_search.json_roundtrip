{
  "functions": [
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "min_guess",
          "type": "int"
        },
        {
          "name": "max_guess",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "min_guess"
          ],
          "dest": "v0",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v1",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "v0",
            "v1"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v2"
          ],
          "dest": "min_plus_1",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "min_plus_1"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "max_guess"
          ],
          "dest": "v5",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v4",
            "v5"
          ],
          "dest": "v6",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "v6"
          ],
          "funcs": [],
          "labels": [
            "then.3",
            "else.3"
          ],
          "op": "br"
        },
        {
          "label": "then.3"
        },
        {
          "args": [
            "min_guess"
          ],
          "dest": "v7",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v7"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "v8",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [],
          "funcs": [],
          "labels": [],
          "op": "ret"
        },
        {
          "label": "else.3"
        },
        {
          "args": [
            "min_guess"
          ],
          "dest": "v9",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "max_guess"
          ],
          "dest": "v10",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v9",
            "v10"
          ],
          "dest": "v11",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v11"
          ],
          "dest": "sum_bounds",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "sum_bounds"
          ],
          "dest": "v12",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v13",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "v12",
            "v13"
          ],
          "dest": "v14",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "v14"
          ],
          "dest": "mid",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "mid"
          ],
          "dest": "v15",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "mid"
          ],
          "dest": "v16",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v15",
            "v16"
          ],
          "dest": "v17",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v17"
          ],
          "dest": "mid_squared",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "mid_squared"
          ],
          "dest": "v19",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v20",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v19",
            "v20"
          ],
          "dest": "v21",
          "funcs": [],
          "labels": [],
          "op": "le",
          "type": "bool"
        },
        {
          "args": [
            "v21"
          ],
          "funcs": [],
          "labels": [
            "then.18",
            "else.18"
          ],
          "op": "br"
        },
        {
          "label": "then.18"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v22",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "mid"
          ],
          "dest": "v23",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "max_guess"
          ],
          "dest": "v24",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v22",
            "v23",
            "v24"
          ],
          "funcs": [
            "main"
          ],
          "labels": [],
          "op": "call"
        },
        {
          "dest": "v25",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "endif.18"
          ],
          "op": "jmp"
        },
        {
          "label": "else.18"
        },
        {
          "args": [
            "x"
          ],
          "dest": "v26",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "min_guess"
          ],
          "dest": "v27",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "mid"
          ],
          "dest": "v28",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v26",
            "v27",
            "v28"
          ],
          "funcs": [
            "main"
          ],
          "labels": [],
          "op": "call"
        },
        {
          "dest": "v29",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "label": "endif.18"
        }
      ],
      "name": "main"
    }
  ]
}
