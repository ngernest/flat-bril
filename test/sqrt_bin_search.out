@main
	op: id   		dest: v0		type: int		args: ["min_guess"]	
	op: const		dest: v1		type: int		value: 1	
	op: add  		dest: v2		type: int		args: ["v0", "v1"]	
	op: id   		dest: min_plus_1		type: int		args: ["v2"]	
	op: id   		dest: v4		type: int		args: ["min_plus_1"]	
	op: id   		dest: v5		type: int		args: ["max_guess"]	
	op: eq   		dest: v6		type: bool		args: ["v4", "v5"]	
	op: br   		args: ["v6"]		labels: ["then.3", "else.3"]
.then.3
	op: id   		dest: v7		type: int		args: ["min_guess"]	
	op: print		args: ["v7"]	
	op: const		dest: v8		type: int		value: 0	
	op: ret  	
.else.3
	op: id   		dest: v9		type: int		args: ["min_guess"]	
	op: id   		dest: v10		type: int		args: ["max_guess"]	
	op: add  		dest: v11		type: int		args: ["v9", "v10"]	
	op: id   		dest: sum_bounds		type: int		args: ["v11"]	
	op: id   		dest: v12		type: int		args: ["sum_bounds"]	
	op: const		dest: v13		type: int		value: 2	
	op: div  		dest: v14		type: int		args: ["v12", "v13"]	
	op: id   		dest: mid		type: int		args: ["v14"]	
	op: id   		dest: v15		type: int		args: ["mid"]	
	op: id   		dest: v16		type: int		args: ["mid"]	
	op: mul  		dest: v17		type: int		args: ["v15", "v16"]	
	op: id   		dest: mid_squared		type: int		args: ["v17"]	
	op: id   		dest: v19		type: int		args: ["mid_squared"]	
	op: id   		dest: v20		type: int		args: ["x"]	
	op: le   		dest: v21		type: bool		args: ["v19", "v20"]	
	op: br   		args: ["v21"]		labels: ["then.18", "else.18"]
.then.18
	op: id   		dest: v22		type: int		args: ["x"]	
	op: id   		dest: v23		type: int		args: ["mid"]	
	op: id   		dest: v24		type: int		args: ["max_guess"]	
	op: call 		args: ["v22", "v23", "v24"]		funcs: ["main"]
	op: const		dest: v25		type: int		value: 0	
	op: jmp  		labels: ["endif.18"]
.else.18
	op: id   		dest: v26		type: int		args: ["x"]	
	op: id   		dest: v27		type: int		args: ["min_guess"]	
	op: id   		dest: v28		type: int		args: ["mid"]	
	op: call 		args: ["v26", "v27", "v28"]		funcs: ["main"]
	op: const		dest: v29		type: int		value: 0	
.endif.18
